base/coreir.jl

```diff
@@ -23,19 +24,64 @@ some elements are known to be constants or a struct whose `Any`-typed field is i
 with `Int` values.
 
 - `typ` indicates the type of the object
-- `fields` holds the lattice elements corresponding to each field of the object
+- `undef` records which fields are possibly undefined
+- `fields` holds the lattice elements corresponding to each defined field of the object
 
-If `typ` is a struct, `fields` represents the fields of the struct that are guaranteed to be
-initialized. For instance, if the length of `fields` of `PartialStruct` representing a
-struct with 4 fields is 3, the 4th field may not be initialized. If the length is 4, all
-fields are guaranteed to be initialized.
+If `typ` is a struct, `undef` represents whether the corresponding field of the struct is guaranteed to be
+initialized. For any defined field (`undef[i] === false`), there is a corresponding `fields` element
+which provides information about the type of the defined field.
 
 If `typ` is a tuple, the last element of `fields` may be `Vararg`. In this case, it is
 guaranteed that the number of elements in the tuple is at least `length(fields)-1`, but the
-exact number of elements is unknown.
+exact number of elements is unknown (`undef` then has a length of `length(fields)-1`).
 """
 Core.PartialStruct
 
+function Core.PartialStruct(typ::Type, undef::BitVector, fields::Vector{Any})
+    @assert length(undef) ≥ length(fields) - (fields[end] === Vararg)
+    for i in 1:datatype_min_ninitialized(typ)
+        undef[i] = false
```

Comment:
I somewhat dislike mutating `undef` here in a place that is quite unexpected, this could be moved to an assert and/or require callsites to do it. However it seemed fairly convenient and more robust to enforce it here.

Reply:
Doing this mutation here isn't ideal. The fact that we need this mutation suggests there's a mistake in how we're handling information at the call site. I think we should change this mutation to an assertion and fix the problem if it's actually happening.